File: src/robinhood_options_mobile/functions/src/news-intelligence.ts
import { onCall, HttpsError } from "firebase-functions/v2/https";
import * as admin from "firebase-admin";
import * as logger from "firebase-functions/logger";

if (admin.apps.length === 0) {
  admin.initializeApp();
}

const db = admin.firestore();

export interface NewsArticle {
  id: string;
  title: string;
  summary: string;
  source: string;
  url: string;
  publishedAt: string;
  sentimentScore: number; // 0 to 100 (50 = neutral)
  sentimentLabel: "Bullish" | "Neutral" | "Bearish";
  impact: "High" | "Medium" | "Low";
  symbols: string[];
}

export interface NewsIntelligence {
  symbol: string;
  overallSentiment: number; // 0-100
  sentimentLabel:
  | "Very Bullish"
  | "Bullish"
  | "Neutral"
  | "Bearish"
  | "Very Bearish";
  headlineSummary: string;
  keyTakeaways: string[];
  bullishCatalysts: string[];
  bearishCatalysts: string[];
  impactRating: "High" | "Medium" | "Low";
  sentimentScoreChange24h: number;
  articles: NewsArticle[];
  updatedAt: string;
}

const BULLISH_KEYWORDS = [
  "beat", "beats", "surpassed", "record", "growth", "upgrade", "upgraded",
  "outperform", "buy", "bullish", "profit", "profitable", "revenue jump",
  "guidance raise", "raised guidance", "partnership", "expansion",
  "dividend increase", "buyback", "fda approval", "approved", "breakthrough",
  "rally", "surges", "soars", "demand surge", "tailwind", "all-time high",
  "strong earnings", "strong quarter",
];

const BEARISH_KEYWORDS = [
  "miss", "misses", "missed", "drop", "drops", "slump", "downgrade",
  "downgraded", "underperform", "sell", "bearish", "loss", "loss widens",
  "revenue fall", "guidance cut", "lowered guidance", "cut guidance",
  "lawsuit", "sued", "investigation", "sec probe", "probe", "regulatory",
  "headwind", "tariff", "inflation", "supply chain delay", "recession",
  "plunges", "tumbles", "debt", "default", "dilution",
];

const HIGH_IMPACT_KEYWORDS = [
  "earnings", "revenue", "guidance", "fda", "merger", "acquisition", "buyout",
  "investigation", "probe", "lawsuit", "antitrust", "dividend", "restructure",
  "ceo",
];

/**
 * Score a single headline/summary text for sentiment and impact.
 * @param {string} text - Title or summary of article.
 * @return {{ score: number, label: string, impact: string }}
 */
export function scoreArticleText(text: string): {
  score: number;
  label: "Bullish" | "Neutral" | "Bearish";
  impact: "High" | "Medium" | "Low";
} {
  const lower = text.toLowerCase();
  let bullCount = 0;
  let bearCount = 0;
  let isHighImpact = false;

  for (const word of BULLISH_KEYWORDS) {
    if (lower.includes(word)) bullCount++;
  }
  for (const word of BEARISH_KEYWORDS) {
    if (lower.includes(word)) bearCount++;
  }
  for (const word of HIGH_IMPACT_KEYWORDS) {
    if (lower.includes(word)) {
      isHighImpact = true;
      break;
    }
  }

  let score = 50;
  if (bullCount > bearCount) {
    score = Math.min(100, 55 + (bullCount - bearCount) * 12);
  } else if (bearCount > bullCount) {
    score = Math.max(0, 45 - (bearCount - bullCount) * 12);
  }

  let label: "Bullish" | "Neutral" | "Bearish" = "Neutral";
  if (score >= 60) label = "Bullish";
  else if (score <= 40) label = "Bearish";

  let impact: "High" | "Medium" | "Low" = "Low";
  if (isHighImpact || Math.abs(score - 50) >= 25) {
    impact = "High";
  } else if (bullCount > 0 || bearCount > 0 || Math.abs(score - 50) >= 10) {
    impact = "Medium";
  }

  return { score, label, impact };
}

/**
 * Analyzes news items for a given symbol and returns NewsIntelligence.
 * @param {any[]} rawArticles - Raw news articles.
 * @param {string} symbol - Ticker symbol.
 * @return {NewsIntelligence} Analyzed intelligence payload.
 */
export function analyzeNewsArticles(
  rawArticles: any[],
  symbol: string
): NewsIntelligence {
  const articles: NewsArticle[] = [];
  const bullishCatalysts: string[] = [];
  const bearishCatalysts: string[] = [];
  const takeaways: string[] = [];

  let totalScore = 0;
  let highImpactCount = 0;

  for (let i = 0; i < rawArticles.length; i++) {
    const raw = rawArticles[i];
    const title = raw.title || raw.headline || "Market Update";
    const summary = raw.summary || raw.description || raw.preview_text || "";
    const source = raw.source || raw.publisher || "Financial News";
    const url = raw.url || raw.link || "";
    const publishedAt =
      raw.published_at ||
      raw.publishedAt ||
      raw.datetime ||
      new Date().toISOString();
    const id = raw.id || `${symbol}-news-${i}-${Date.now()}`;

    const textToEvaluate = `${title} ${summary}`;
    const { score, label, impact } = scoreArticleText(textToEvaluate);

    if (impact === "High") highImpactCount++;
    totalScore += score;

    if (label === "Bullish" && bullishCatalysts.length < 4) {
      bullishCatalysts.push(title);
    } else if (label === "Bearish" && bearishCatalysts.length < 4) {
      bearishCatalysts.push(title);
    }

    articles.push({
      id: String(id),
      title,
      summary,
      source,
      url,
      publishedAt,
      sentimentScore: score,
      sentimentLabel: label,
      impact,
      symbols: raw.symbols || [symbol],
    });
  }

  const overallSentiment =
    articles.length > 0 ? Math.round(totalScore / articles.length) : 50;

  let sentimentLabel:
    | "Very Bullish"
    | "Bullish"
    | "Neutral"
    | "Bearish"
    | "Very Bearish" = "Neutral";
  if (overallSentiment >= 75) sentimentLabel = "Very Bullish";
  else if (overallSentiment >= 60) sentimentLabel = "Bullish";
  else if (overallSentiment <= 25) sentimentLabel = "Very Bearish";
  else if (overallSentiment <= 40) sentimentLabel = "Bearish";

  let impactRating: "High" | "Medium" | "Low" = "Low";
  if (highImpactCount >= 2 || Math.abs(overallSentiment - 50) >= 20) {
    impactRating = "High";
  } else if (highImpactCount >= 1 || articles.length >= 3) {
    impactRating = "Medium";
  }

  // Generate headline summary
  let headlineSummary = "";
  if (articles.length === 0) {
    headlineSummary = `No recent news catalysts reported for ${symbol}. ` +
      "Market sentiment remains baseline neutral.";
    takeaways.push("Trading volume and technical factors are primary drivers.");
  } else {
    headlineSummary = `${symbol} news sentiment is ` +
      `${sentimentLabel.toLowerCase()} (${overallSentiment}/100) ` +
      `driven by ${articles.length} recent news events.`;

    if (bullishCatalysts.length > 0) {
      takeaways.push(`Key positive driver: ${bullishCatalysts[0]}`);
    }
    if (bearishCatalysts.length > 0) {
      takeaways.push(`Key risk factor: ${bearishCatalysts[0]}`);
    }
    if (articles.length === 0) {
      takeaways.push("No recent catalysts; trading on technicals.");
    }
  }

  return {
    symbol: symbol.toUpperCase(),
    overallSentiment,
    sentimentLabel,
    headlineSummary,
    keyTakeaways: takeaways,
    bullishCatalysts,
    bearishCatalysts,
    impactRating,
    sentimentScoreChange24h: 0,
    articles,
    updatedAt: new Date().toISOString(),
  };
}

/**
 * Callable Firebase Function to retrieve news intelligence for a symbol.
 */
export const getNewsIntelligence = onCall({ cors: true }, async (request) => {
  const data = request.data || {};
  const symbol = (data.symbol || "").toUpperCase();
  const providedArticles = data.articles || [];

  if (!symbol) {
    throw new HttpsError("invalid-argument", "Symbol must be provided.");
  }

  logger.info(`getNewsIntelligence requested for ${symbol}`, {
    articlesCount: providedArticles.length,
  });

  try {
    // If articles passed directly, analyze them
    if (Array.isArray(providedArticles) && providedArticles.length > 0) {
      return analyzeNewsArticles(providedArticles, symbol);
    }

    // Check cached news intelligence in Firestore if recent (< 15 min)
    const cachedDoc = await db.collection("instrument_news").doc(symbol).get();
    if (cachedDoc.exists && data.refresh !== true) {
      const cachedData = cachedDoc.data() as NewsIntelligence;
      const updatedAtMs = new Date(cachedData.updatedAt || 0).getTime();
      if (Date.now() - updatedAtMs < 15 * 60 * 1000) {
        return cachedData;
      }
    }

    // Fallback: analyze whatever news items exist or baseline
    const intelligence = analyzeNewsArticles([], symbol);
    await db
      .collection("instrument_news")
      .doc(symbol)
      .set(intelligence, { merge: true });
    return intelligence;
  } catch (error: any) {
    logger.error(`Error in getNewsIntelligence for ${symbol}:`, error);
    throw new HttpsError(
      "internal",
      `Failed to get news intelligence: ${error.message}`
    );
  }
});

/**
 * Callable Firebase Function for multiple watchlist symbols.
 */
export const getWatchlistNewsIntelligence = onCall(
  { cors: true },
  async (request) => {
    const data = request.data || {};
    const symbols = (data.symbols || []) as string[];

    if (!Array.isArray(symbols) || symbols.length === 0) {
      throw new HttpsError(
        "invalid-argument",
        "Symbols array must be provided."
      );
    }

    logger.info(
      `getWatchlistNewsIntelligence for ${symbols.length} symbols`
    );

    const results: Record<string, NewsIntelligence> = {};

    try {
      const limitedSymbols = symbols.slice(0, 25).map((s) => s.toUpperCase());
      const snapshots = await Promise.all(
        limitedSymbols.map((s) =>
          db.collection("instrument_news").doc(s).get()
        )
      );

      for (let i = 0; i < limitedSymbols.length; i++) {
        const sym = limitedSymbols[i];
        const snap = snapshots[i];
        if (snap.exists) {
          results[sym] = snap.data() as NewsIntelligence;
        } else {
          results[sym] = analyzeNewsArticles([], sym);
        }
      }

      return results;
    } catch (error: any) {
      logger.error("Error in getWatchlistNewsIntelligence:", error);
      throw new HttpsError(
        "internal",
        `Failed to get watchlist news: ${error.message}`
      );
    }
  }
);

File: src/robinhood_options_mobile/lib/model/news_intelligence.dart
import 'package:flutter/material.dart';

enum NewsSentimentLabel {
  veryBullish,
  bullish,
  neutral,
  bearish,
  veryBearish;

  String get displayName {
    switch (this) {
      case NewsSentimentLabel.veryBullish:
        return 'Very Bullish';
      case NewsSentimentLabel.bullish:
        return 'Bullish';
      case NewsSentimentLabel.neutral:
        return 'Neutral';
      case NewsSentimentLabel.bearish:
        return 'Bearish';
      case NewsSentimentLabel.veryBearish:
        return 'Very Bearish';
    }
  }

  Color get color {
    switch (this) {
      case NewsSentimentLabel.veryBullish:
        return Colors.green.shade700;
      case NewsSentimentLabel.bullish:
        return Colors.green;
      case NewsSentimentLabel.neutral:
        return Colors.grey;
      case NewsSentimentLabel.bearish:
        return Colors.orange.shade800;
      case NewsSentimentLabel.veryBearish:
        return Colors.red;
    }
  }

  static NewsSentimentLabel fromString(String? value) {
    switch (value?.toLowerCase()) {
      case 'very bullish':
      case 'verybullish':
        return NewsSentimentLabel.veryBullish;
      case 'bullish':
        return NewsSentimentLabel.bullish;
      case 'very bearish':
      case 'verybearish':
        return NewsSentimentLabel.veryBearish;
      case 'bearish':
        return NewsSentimentLabel.bearish;
      default:
        return NewsSentimentLabel.neutral;
    }
  }
}

enum NewsImpact {
  high,
  medium,
  low;

  String get displayName {
    switch (this) {
      case NewsImpact.high:
        return 'High Impact';
      case NewsImpact.medium:
        return 'Medium Impact';
      case NewsImpact.low:
        return 'Low Impact';
    }
  }

  Color get color {
    switch (this) {
      case NewsImpact.high:
        return Colors.purpleAccent;
      case NewsImpact.medium:
        return Colors.blueAccent;
      case NewsImpact.low:
        return Colors.grey;
    }
  }

  static NewsImpact fromString(String? value) {
    switch (value?.toLowerCase()) {
      case 'high':
        return NewsImpact.high;
      case 'medium':
        return NewsImpact.medium;
      default:
        return NewsImpact.low;
    }
  }
}

class NewsArticleItem {
  final String id;
  final String title;
  final String summary;
  final String source;
  final String url;
  final DateTime publishedAt;
  final double sentimentScore; // 0 to 100
  final NewsSentimentLabel sentimentLabel;
  final NewsImpact impact;
  final List<String> symbols;

  const NewsArticleItem({
    required this.id,
    required this.title,
    required this.summary,
    required this.source,
    required this.url,
    required this.publishedAt,
    required this.sentimentScore,
    required this.sentimentLabel,
    required this.impact,
    this.symbols = const [],
  });

  factory NewsArticleItem.fromMap(Map<String, dynamic> map) {
    return NewsArticleItem(
      id: map['id']?.toString() ?? '',
      title: map['title']?.toString() ?? '',
      summary:
          map['summary']?.toString() ?? map['description']?.toString() ?? '',
      source:
          map['source']?.toString() ?? map['publisher']?.toString() ?? 'News',
      url: map['url']?.toString() ?? '',
      publishedAt: map['publishedAt'] != null
          ? DateTime.tryParse(map['publishedAt'].toString()) ?? DateTime.now()
          : (map['published_at'] != null
              ? DateTime.tryParse(map['published_at'].toString()) ??
                  DateTime.now()
              : DateTime.now()),
      sentimentScore: (map['sentimentScore'] as num?)?.toDouble() ?? 50.0,
      sentimentLabel:
          NewsSentimentLabel.fromString(map['sentimentLabel']?.toString()),
      impact: NewsImpact.fromString(map['impact']?.toString()),
      symbols: (map['symbols'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'title': title,
      'summary': summary,
      'source': source,
      'url': url,
      'publishedAt': publishedAt.toIso8601String(),
      'sentimentScore': sentimentScore,
      'sentimentLabel': sentimentLabel.displayName,
      'impact': impact.name,
      'symbols': symbols,
    };
  }
}

class EventImpactPrediction {
  final String direction;
  final double expectedMovePercent;
  final int confidence;
  final String horizon;
  final List<String> drivers;

  const EventImpactPrediction({
    required this.direction,
    required this.expectedMovePercent,
    required this.confidence,
    required this.horizon,
    this.drivers = const [],
  });

  factory EventImpactPrediction.fromMap(Map<String, dynamic>? map) {
    return EventImpactPrediction(
      direction: map?['direction']?.toString() ?? 'Neutral',
      expectedMovePercent:
          (map?['expectedMovePercent'] as num?)?.toDouble() ?? 0.0,
      confidence: (map?['confidence'] as num?)?.toInt() ?? 15,
      horizon: map?['horizon']?.toString() ?? '1-2 weeks',
      drivers: (map?['drivers'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }
}

class NewsIntelligence {
  final String symbol;
  final double overallSentiment; // 0-100
  final NewsSentimentLabel sentimentLabel;
  final String headlineSummary;
  final List<String> keyTakeaways;
  final List<String> bullishCatalysts;
  final List<String> bearishCatalysts;
  final NewsImpact impactRating;
  final double sentimentScoreChange24h;
  final List<NewsArticleItem> articles;
  final EventImpactPrediction eventImpactPrediction;
  final DateTime updatedAt;

  const NewsIntelligence({
    required this.symbol,
    required this.overallSentiment,
    required this.sentimentLabel,
    required this.headlineSummary,
    this.keyTakeaways = const [],
    this.bullishCatalysts = const [],
    this.bearishCatalysts = const [],
    required this.impactRating,
    this.sentimentScoreChange24h = 0.0,
    this.articles = const [],
    this.eventImpactPrediction = const EventImpactPrediction(
      direction: 'Neutral',
      expectedMovePercent: 0,
      confidence: 15,
      horizon: '1-2 weeks',
    ),
    required this.updatedAt,
  });

  factory NewsIntelligence.fromMap(Map<String, dynamic> map) {
    final rawArticles = map['articles'] as List<dynamic>? ?? [];
    return NewsIntelligence(
      symbol: map['symbol']?.toString() ?? '',
      overallSentiment: (map['overallSentiment'] as num?)?.toDouble() ?? 50.0,
      sentimentLabel:
          NewsSentimentLabel.fromString(map['sentimentLabel']?.toString()),
      headlineSummary: map['headlineSummary']?.toString() ?? '',
      keyTakeaways: (map['keyTakeaways'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      bullishCatalysts: (map['bullishCatalysts'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      bearishCatalysts: (map['bearishCatalysts'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      impactRating: NewsImpact.fromString(map['impactRating']?.toString()),
      sentimentScoreChange24h:
          (map['sentimentScoreChange24h'] as num?)?.toDouble() ?? 0.0,
      articles: rawArticles
          .whereType<Map>()
          .map((a) => NewsArticleItem.fromMap(Map<String, dynamic>.from(a)))
          .toList(),
      eventImpactPrediction: EventImpactPrediction.fromMap(
          map['eventImpactPrediction'] as Map<String, dynamic>?),
      updatedAt: map['updatedAt'] != null
          ? DateTime.tryParse(map['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'symbol': symbol,
      'overallSentiment': overallSentiment,
      'sentimentLabel': sentimentLabel.displayName,
      'headlineSummary': headlineSummary,
      'keyTakeaways': keyTakeaways,
      'bullishCatalysts': bullishCatalysts,
      'bearishCatalysts': bearishCatalysts,
      'impactRating': impactRating.name,
      'sentimentScoreChange24h': sentimentScoreChange24h,
      'articles': articles.map((a) => a.toMap()).toList(),
      'eventImpactPrediction': {
        'direction': eventImpactPrediction.direction,
        'expectedMovePercent': eventImpactPrediction.expectedMovePercent,
        'confidence': eventImpactPrediction.confidence,
        'horizon': eventImpactPrediction.horizon,
        'drivers': eventImpactPrediction.drivers,
      },
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}
